cmd/ttn-lw-stack/commands/ns_db.go

```diff
@@ -92,377 +90,64 @@ var (
-						}
-						return true, nil
-					}); err != nil {
-						logger.WithError(err).Error("Failed to scan current session field key")
-						return true, nil
-					}
-					if migratedSubkeys == 0 {
-						return true, nil
-					}
+			logger.Info("Connecting to Redis database...")
+			uplinkClient := NewNetworkServerApplicationUplinkQueueRedis(config)
+			defer uplinkClient.Close()
 
-				case addrRegexp3_11_PendingFields.MatchString(k):
-					var migratedSubkeys uint64
-					if err := ttnredis.RangeRedisHMap(ctx, cl, k, "*", 1, func(uid string, s string) (bool, error) {
-						logger := logger.WithField("uid", uid)
-						if err := cl.Watch(ctx, func(tx *redis.Tx) error {
-							if err := msgpack.Unmarshal([]byte(s), &nsredis.UplinkMatchPendingSession{}); err == nil {
-								return nil
-							}
-							stored := &struct {
-								FNwkSIntKey    KeyEnvelope3_11Snapshot
-								LoRaWANVersion ttnpb.MACVersion
-							}{}
-							if err := msgpack.Unmarshal([]byte(s), stored); err != nil {
-								logger.WithError(err).Error("Failed to unmarshal legacy pending session fields")
-								return err
-							}
-							var key *types.AES128Key
-							if len(stored.FNwkSIntKey.Key) > 0 {
-								key = &types.AES128Key{}
-								copy(key[:], stored.FNwkSIntKey.Key)
-							}
-							b, err := msgpack.Marshal(&nsredis.UplinkMatchSession{
-								FNwkSIntKey: &ttnpb.KeyEnvelope{
-									Key:          key.Bytes(),
-									KekLabel:     stored.FNwkSIntKey.KEKLabel,
-									EncryptedKey: stored.FNwkSIntKey.EncryptedKey,
-								},
-								LoRaWANVersion: stored.LoRaWANVersion,
-							})
-							if err != nil {
-								logger.WithError(err).Error("Failed to marshal pending session fields")
-								return nil
-							}
-							if err := tx.HSet(ctx, k, uid, string(b)).Err(); err != nil {
-								logger.WithError(err).Error("Failed to set pending session fields")
-								return err
-							}
-							migratedSubkeys++
-							return nil
-						}, k); err != nil {
-							logger.WithError(err).Error("Transaction failed")
-						}
-						return true, nil
-					}); err != nil {
-						logger.WithError(err).Error("Failed to scan pending session field key")
-						return true, nil
-					}
-					if migratedSubkeys == 0 {
-						return true, nil
-					}
+			var removed uint64
+			defer func() { logger.Debugf("%d keys removed", removed) }()
 
-				case uidRegexp.MatchString(k),
-					euiRegexp.MatchString(k),
-					addrRegexp3_11_Current.MatchString(k):
-					logger.Debug("Skip valid key")
-					return true, nil
+			uidLastInvalidationKey := regexp.MustCompile(uplinkClient.Key("uid", ".*", "last-invalidation$"))
 
-				default:
-					d, err := cl.TTL(ctx, k).Result()
-					if err != nil {
-						logger.WithError(err).Error("Failed to determine TTL of unmatched key")
+			pipeliner := uplinkClient.Pipeline()
+			err := ttnredis.RangeRedisKeys(ctx, uplinkClient, uplinkClient.Key("*"), ttnredis.DefaultRangeCount,
```

Comment:
Using `Key("*")` is very wide - you will iterate through the whole uplinks space. Wouldn't something like `Key("uid", "*", "last-invalidations")` be better ?

Reply:
True. Updated it and removed the RegEx checking as it was not needed anymore because of the specific uplink space.